File: domain/src/main/java/ch/threema/domain/onprem/OnPremConfigParser.java
/*  _____ _
 * |_   _| |_  _ _ ___ ___ _ __  __ _
 *   | | | ' \| '_/ -_) -_) '  \/ _` |_
 *   |_| |_||_|_| \___\___|_|_|_\__,_(_)
 *
 * Threema for Android
 * Copyright (c) 2020-2024 Threema GmbH
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License, version 3,
 * as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package ch.threema.domain.onprem;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;
import org.slf4j.Logger;

import java.io.IOException;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;

import ch.threema.base.utils.Base64;
import ch.threema.base.utils.LoggingUtil;

public class OnPremConfigParser {

    private static final Logger logger = LoggingUtil.getThreemaLogger("OnPremConfigParser");

    private static final int MIN_REFRESH_VALUE_S = 1800;

    public OnPremConfig parse(JSONObject obj) throws IOException, JSONException, ParseException, LicenseExpiredException {
        int refreshValue = obj.getInt("refresh");
        if (refreshValue < MIN_REFRESH_VALUE_S) {
            logger.warn("Invalid refresh value provided: {}; using {} as fallback", refreshValue, MIN_REFRESH_VALUE_S);
            refreshValue = MIN_REFRESH_VALUE_S;
        }

        return new OnPremConfig(
            System.currentTimeMillis() + refreshValue * 1000L,
            this.parseLicense(obj.getJSONObject("license")),
            this.parseChatConfig(obj.getJSONObject("chat")),
            this.parseDirectoryConfig(obj.getJSONObject("directory")),
            this.parseBlobConfig(obj.getJSONObject("blob")),
            this.parseWorkConfig(obj.getJSONObject("work")),
            this.parseAvatarConfig(obj.getJSONObject("avatar")),
            this.parseSafeConfig(obj.getJSONObject("safe")),
            this.parseWebConfig(obj.optJSONObject("web")),
            this.parseMediatorConfig(obj.optJSONObject("mediator")));
    }

    private OnPremLicense parseLicense(JSONObject obj) throws JSONException, ParseException, LicenseExpiredException {
        SimpleDateFormat ymdFormat = new SimpleDateFormat("yyyy-MM-dd");
        Date expires = ymdFormat.parse(obj.getString("expires"));

        // Add one day as expiration is defined as "not valid after this date"
        expires.setTime(expires.getTime() + 86400000);

        // Check that license has not expired
        if (expires.before(new Date())) {
            throw new LicenseExpiredException("License has expired");
        }

        return new OnPremLicense(obj.getString("id"), expires, obj.getInt("count"));
    }

    private OnPremConfigChat parseChatConfig(JSONObject obj) throws JSONException, IOException {
        JSONArray portsArray = obj.getJSONArray("ports");
        int[] ports = new int[portsArray.length()];
        for (int i = 0; i < ports.length; i++) {
            ports[i] = portsArray.getInt(i);
        }
        return new OnPremConfigChat(obj.getString("hostname"), ports, Base64.decode(obj.getString("publicKey")));
    }

    private OnPremConfigDirectory parseDirectoryConfig(JSONObject obj) throws JSONException {
        return new OnPremConfigDirectory(obj.getString("url"));
    }

    private OnPremConfigBlob parseBlobConfig(JSONObject obj) throws JSONException {
        return new OnPremConfigBlob(obj.getString("uploadUrl"), obj.getString("downloadUrl"), obj.getString("doneUrl"));
    }

    private OnPremConfigAvatar parseAvatarConfig(JSONObject obj) throws JSONException {
        return new OnPremConfigAvatar(obj.getString("url"));
    }

    private OnPremConfigSafe parseSafeConfig(JSONObject obj) throws JSONException {
        return new OnPremConfigSafe(obj.getString("url"));
    }

    private OnPremConfigWork parseWorkConfig(JSONObject obj) throws JSONException {
        return new OnPremConfigWork(obj.getString("url"));
    }

    private OnPremConfigWeb parseWebConfig(JSONObject obj) throws JSONException {
        if (obj == null) {
            return null;
        }
        return new OnPremConfigWeb(obj.getString("url"), obj.optString("overrideSaltyRtcHost"), obj.optInt("overrideSaltyRtcPort"));
    }

    private OnPremConfigMediator parseMediatorConfig(JSONObject obj) throws JSONException {
        if (obj == null) {
            return null;
        }
        return new OnPremConfigMediator(obj.getString("url"), parseBlobConfig(obj.getJSONObject("blob")));
    }
}

File: domain/src/main/java/ch/threema/domain/onprem/OnPremConfigParser.kt
package ch.threema.domain.onprem

import ch.threema.base.utils.Base64
import ch.threema.base.utils.getThreemaLogger
import ch.threema.common.plus
import ch.threema.common.toIntArray
import ch.threema.common.toJSONObjectList
import ch.threema.domain.protocol.urls.BlobUrl
import ch.threema.domain.protocol.urls.MapPoiAroundUrl
import ch.threema.domain.protocol.urls.MapPoiNamesUrl
import ch.threema.domain.protocol.urls.MediatorUrl
import java.io.IOException
import java.text.ParseException
import java.time.Instant
import java.time.LocalDate
import java.time.ZoneOffset
import kotlin.time.Duration
import kotlin.time.Duration.Companion.minutes
import kotlin.time.Duration.Companion.seconds
import org.json.JSONException
import org.json.JSONObject

private val logger = getThreemaLogger("OnPremConfigParser")

class OnPremConfigParser {
    @Throws(OnPremConfigParseException::class)
    fun parse(obj: JSONObject, createdAt: Instant): OnPremConfig =
        try {
            OnPremConfig(
                validUntil = createdAt + getRefreshDuration(obj),
                license = parseLicense(obj.getJSONObject("license")),
                domains = parseDomains(obj.optJSONObject("domains")),
                chat = parseChatConfig(obj.getJSONObject("chat")),
                directory = parseDirectoryConfig(obj.getJSONObject("directory")),
                blob = parseBlobConfig(obj.getJSONObject("blob")),
                work = parseWorkConfig(obj.getJSONObject("work")),
                avatar = parseAvatarConfig(obj.getJSONObject("avatar")),
                safe = parseSafeConfig(obj.getJSONObject("safe")),
                web = parseWebConfig(obj.optJSONObject("web")),
                mediator = parseMediatorConfig(obj.optJSONObject("mediator")),
                maps = parseMapConfig(obj.optJSONObject("maps")),
            )
        } catch (e: IOException) {
            throw OnPremConfigParseException(e)
        } catch (e: JSONException) {
            throw OnPremConfigParseException(e)
        } catch (e: ParseException) {
            throw OnPremConfigParseException(e)
        }

    private fun getRefreshDuration(obj: JSONObject): Duration {
        val refreshDuration = obj.getInt("refresh").seconds
        if (refreshDuration < minRefreshDuration) {
            logger.warn("Invalid refresh duration provided: {}; using {} as fallback", refreshDuration, minRefreshDuration)
            return minRefreshDuration
        }
        return refreshDuration
    }

    private fun parseLicense(obj: JSONObject): OnPremLicense =
        OnPremLicense(
            id = obj.getString("id"),
            expires = Instant.from(
                LocalDate.parse(obj.getString("expires"))
                    .atStartOfDay()
                    .plusDays(1)
                    .atOffset(ZoneOffset.UTC),
            ),
            count = obj.getInt("count"),
        )

    private fun parseDomains(obj: JSONObject?): OnPremConfigDomains? {
        val rules = obj?.getJSONArray("rules") ?: return null
        return OnPremConfigDomains(
            rules = rules
                .toJSONObjectList()
                .mapNotNull { ruleObject ->
                    val matchModeString = ruleObject.getString("matchMode")
                    val matchMode = OnPremConfigDomainRuleMatchMode.fromStringOrNull(matchModeString)
                    if (matchMode == null) {
                        logger.warn("Unknown match mode '{}' found, ignoring domain rule", matchModeString)
                        return@mapNotNull null
                    }
                    OnPremConfigDomainRule(
                        fqdn = ruleObject.getString("fqdn"),
                        matchMode = matchMode,
                        spkis = ruleObject.optJSONArray("spkis")
                            ?.toJSONObjectList()
                            ?.mapNotNull { spkiObject ->
                                val algorithmString = spkiObject.getString("algorithm")
                                val algorithm = OnPremConfigDomainRuleSpkiAlgorithm.fromStringOrNull(algorithmString)
                                if (algorithm != null) {
                                    OnPremConfigDomainRuleSpki(
                                        value = spkiObject.getString("value"),
                                        algorithm = algorithm,
                                    )
                                } else {
                                    logger.warn("Unknown algorithm '{}' found, ignoring domain rule spki", algorithmString)
                                    null
                                }
                            },
                    )
                },
        )
    }

    private fun parseChatConfig(obj: JSONObject): OnPremConfigChat =
        OnPremConfigChat(
            hostname = obj.getString("hostname"),
            ports = obj.getJSONArray("ports").toIntArray(),
            publicKey = Base64.decode(obj.getString("publicKey")),
        )

    private fun parseDirectoryConfig(obj: JSONObject): OnPremConfigDirectory =
        OnPremConfigDirectory(obj.getString("url"))

    private fun parseBlobConfig(obj: JSONObject): OnPremConfigBlob =
        OnPremConfigBlob(
            uploadUrl = obj.getString("uploadUrl"),
            downloadUrl = BlobUrl(obj.getString("downloadUrl")),
            doneUrl = BlobUrl(obj.getString("doneUrl")),
        )

    private fun parseAvatarConfig(obj: JSONObject): OnPremConfigAvatar =
        OnPremConfigAvatar(url = obj.getString("url"))

    private fun parseSafeConfig(obj: JSONObject): OnPremConfigSafe =
        OnPremConfigSafe(url = obj.getString("url"))

    private fun parseWorkConfig(obj: JSONObject): OnPremConfigWork =
        OnPremConfigWork(url = obj.getString("url"))

    private fun parseWebConfig(obj: JSONObject?): OnPremConfigWeb? =
        obj?.let {
            OnPremConfigWeb(
                url = obj.getString("url"),
                overrideSaltyRtcHost = obj.optString("overrideSaltyRtcHost"),
                overrideSaltyRtcPort = obj.optInt("overrideSaltyRtcPort"),
            )
        }

    private fun parseMediatorConfig(obj: JSONObject?): OnPremConfigMediator? =
        obj?.let {
            OnPremConfigMediator(
                url = MediatorUrl(obj.getString("url")),
                blob = parseBlobConfig(obj.getJSONObject("blob")),
            )
        }

    private fun parseMapConfig(obj: JSONObject?): OnPremConfigMaps? =
        obj?.let {
            OnPremConfigMaps(
                styleUrl = obj.getString("styleUrl"),
                poiNamesUrl = MapPoiNamesUrl(obj.getString("poiNamesUrl")),
                poiAroundUrl = MapPoiAroundUrl(obj.getString("poiAroundUrl")),
            )
        }

    companion object {
        private val minRefreshDuration = 30.minutes
    }
}
